File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            ans[i] = 1;
        }

        int accumulated_product = 1;
        for (int i = 0; i < n - 1; i++) {
            accumulated_product *= nums[i];
            ans[i + 1] *= accumulated_product;
        }
        
        accumulated_product = 1;
        for (int i = n - 1; i > 0; i--) {
            accumulated_product *= nums[i];
            ans[i - 1] *= accumulated_product;
        }

        return ans;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        ans = [1] * len(nums)
        prefix_product, suffix_product = 1, 1

        for i in range(len(nums)):
            ans[i] *= prefix_product
            prefix_product *= nums[i]
            ans[-1 - i] *= suffix_product
            suffix_product *= nums[-1 - i]

        return ans